misc/pragmas.md

```diff
@@ -0,0 +1,550 @@
+long complicated annotations.
+
+
+## Worked examples
+
+*This section should contain, for some or all of the use cases, fully
+worked examples showing simple grammars that use the annotations in
+question.*
+
+
+### Namespace declarations.
+
+Using pragmas to specify that all or some elements of the XML returned
+by an ixml processor should go into a specified namespace.
+
+For example, let's assume that we want some elements to go into one
+namespace, some into another, and some into none, and that namespace
+bindings should remain constant throughout the grammar (so: no
+changing the default namespace in the middle of the document).
+
+We define the namespace 'http://example.com/ixml-namespaces" as
+providing for namespace bindings, and we adopt the convention that a
+prefix `ppp` is bound to a namespace name `nnn` if a pragma of the form
+````
+    [@nsd:ppp nnn]
+````
+
+is encountered in a context where the prefix *nsd* is bound to
+"'http://example.com/ixml-namespaces".
+
+To bootstrap the process, we adopt the principle that a 
+pragma of the form
+````
+    [@ns:ns http://example.com/ixml-namespaces]
+````
+
+binds the prefix `ns` to that namespace. (Alternatively, we could
+follow the example of XML namespaces and reserve some string for
+namespace bindings, but this example is trying to show a way to do
+this that does not rely on a decree in the spec but could in principle
+be handled as an extension.)
+
+So a grammar in which some elements are in the *x* namespace, some in
+the *y* namespace, and some in no namespace at all, might be
+
+````
+    [@nsd:nsd http://example.com/ixml-namespaces]
+    [@nsd:x http://example.org/NS/existential]
+    [@nsd:y http://example.com/NS/yoyo]
+    x:sentence: x:a, ' ', y:b?, '.  ', c.
+	x:a: 'Speed'.
+	y:b: 'kills'.
+	c:  'It really does.'.
+````
+The XML representation of the grammar might be:
+
+````
+    <ixml>
+      <rule name="x:sentence"
+        nsd:nsd="http://example.com/ixml-namespaces"
+        nsd:x="http://example.org/NS/existential"
+        nsd:y="http://example.com/NS/yoyo"
+        >
+        <alt>
+          <nonterminal name="x:a"/>
+          <literal sstring=" "/>
+          <option>
+            <nonterminal name="y:b"/>
+          </option>
+          <literal sstring=".  "/>
+          <nonterminal name="c"/>
+        </alt>
+      </rule>
+      <rule name="x:a">
+        <alt>
+          <literal sstring="Speed"/>
+        </alt>
+      </rule>
+      <rule name="y:b">
+        <alt>
+          <literal sstring="kills"/>
+          </alt>
+      </rule>
+      <rule name="c">
+        <alt>
+          <literal sstring="It really does."/>
+        </alt> 
+      </rule>
+    </ixml>
+````
+
+*We have another bootstrapping problem here: because the prefix*
+`nsd` *is not declared, the XML shown is not namespace-well-formed. We
+need some way of getting it declared, preferably other than by fiat,
+but I don't currently see an alternative to fiat.*
```

Comment:
1. The resulting XML is only not well-formed if the parser understands pragma.
2. I think we have already agreed above that a pragma-aware parser must understand the namespace pragma, and must recognise the form `@ns:ns http://example.com/ixml-namespaces` (or whichever URI ends up being specified)
3. I don't think it's a problematic next step to say that such a pragma must result in an `xmlns:ns="http://example.com/ixml-namespaces"` declaration on the xml representation
4. I DO think that we need to explicitly state how that form is recognised, i.e. that the namespace and local name must be identical, that the pragma is in an attribute form (?), and that the value of the pragma is the given URL.